base/system/userinit/userinit.c

```diff
@@ -552,22 +565,101 @@ NotifyLogon(VOID)
 static BOOL
 StartInstaller(VOID)
 {
-    WCHAR Shell[MAX_PATH];
+    SYSTEM_INFO SystemInfo;
+    PWSTR ptr;
+    WCHAR Installer[MAX_PATH];
     WCHAR szMsg[RC_STRING_MAX_SIZE];
 
-    if (GetWindowsDirectoryW(Shell, ARRAYSIZE(Shell) - 12))
-        wcscat(Shell, L"\\reactos.exe");
+    /*
+     * First, check whether we can start the installer from either the current
+     * ReactOS installation directory, or from our current directory.
+     */
+    *Installer = UNICODE_NULL;
+    /* Alternatively one can use SharedUserData->NtSystemRoot */
+    if (GetSystemWindowsDirectoryW(Installer, ARRAYSIZE(Installer) - 12))
+        wcscat(Installer, L"\\");
+    wcscat(Installer, L"reactos.exe");
+
+    if (StartProcess(Installer))
+        return TRUE;
+
+    ERR("Failed to start the installer: '%s', trying alternative.\n", debugstr_w(Installer));
+
+    /*
+     * We failed. Try using the default drive, under the directory whose
+     * name corresponds to the currently-runnning CPU architecture.
+     */
+    GetSystemInfo(&SystemInfo);
+
+    *Installer = UNICODE_NULL;
+    /* Alternatively one can use SharedUserData->NtSystemRoot */
+    GetSystemWindowsDirectoryW(Installer, ARRAYSIZE(Installer) - 12);
+    ptr = wcschr(Installer, L'\\');
+    if (ptr)
+        *++ptr = UNICODE_NULL;
     else
-        wcscpy(Shell, L"reactos.exe");
+        *Installer = UNICODE_NULL;
 
-    if (!TryToStartShell(Shell))
+    /* Append the corresponding CPU architecture */
+    switch (SystemInfo.wProcessorArchitecture)
```

Comment:
Can this be combined with ExpandInstallerPath somehow?

Reply:
ExpandInstallerPath is in WELCOME, not in USERINIT, so I'm not sure what I mean by this.